fem/field_interpolant.hpp

```diff
@@ -0,0 +1,191 @@
+   virtual void FullReset()
+   {
+      L2->Update();
+      L2->Assemble();
+   }
+   virtual void SetupCG(double rel_tol = 1e-15, double abs_tol = 0.0,
+                        int print_level = 0, int max_iter = 2000)
+   {
+      cg = new CGSolver();
+      cg->SetPrintLevel(print_level);
+      cg->SetMaxIter(max_iter);
+      cg->SetRelTol(rel_tol);
+      cg->SetAbsTol(abs_tol);
+   }
+   ~FieldInterpolant()
+   {
+      delete L2;
+      delete cg;
+   }
+};
+
+#ifdef MFEM_USE_MPI
+class ParFieldInterpolant : public FieldInterpolant
+{
+protected:
+   ParBilinearForm *ParL2;
+public:
+   // The ParFiniteElementSpace passed into here should have a vdim set to 1 in order for the
+   // MassIntegrator to work properly if the ProjectQuadratureCoefficient method is used with
+   // a VectorQuadratureFunctionCoefficient.
+   ParFieldInterpolant(ParFiniteElementSpace *pfes) : FieldInterpolant(pfes)
+   {
+      ParL2 = new ParBilinearForm(pfes);
+
+      const FiniteElement &el = *pfes->GetFE(0);
+      const IntegrationRule *ir = &(IntRules.Get(el.GetGeomType(),
+                                                 2 * el.GetOrder() + 1));
+
+      ParL2->AddDomainIntegrator(new MassIntegrator(ir));
+      ParL2->Assemble();
+   }
+   // This function takes a vector quadrature function coefficient and projects it onto a GridFunction of the same space as vector
+   // quadrature function coefficient.
+   void ProjectQuadratureCoefficient(ParGridFunction &gf,
+                                     VectorQuadratureFunctionCoefficient &vqfc,
+                                     ParFiniteElementSpace &fes);
+   // This function takes a quadrature function coefficient and projects it onto a GridFunction of the same space as
+   // quadrature function coefficient.
+   void ProjectQuadratureCoefficient(ParGridFunction &gf,
+                                     QuadratureFunctionCoefficient &qfc,
+                                     ParFiniteElementSpace &fes);
+   // Tells the internal bilinearform needs to be reset in order to reset the sparse matrix
+   virtual void FullReset() override
+   {
+      FieldInterpolant::FullReset();
+      ParL2->Update();
+      ParL2->Assemble();
+   }
+   using FieldInterpolant::SetupCG;
+   // Setup the CG solver with an MPI communicator
+   virtual void SetupCG(MPI_Comm _comm, double rel_tol = 1e-15,
+                        double abs_tol = 0.0,
+                        int print_level = 0, int max_iter = 2000)
+   {
+      cg = new CGSolver(_comm);
+      cg->SetPrintLevel(print_level);
+      cg->SetMaxIter(max_iter);
+      cg->SetRelTol(rel_tol);
+      cg->SetAbsTol(abs_tol);
+   }
+
+   ~ParFieldInterpolant()
+   {
+      delete ParL2;
+   }
+};
+#endif
+
+class VectorQuadratureIntegrator : public LinearFormIntegrator
+{
+private:
+   VectorQuadratureFunctionCoefficient &vqfc;
+public:
+   VectorQuadratureIntegrator(VectorQuadratureFunctionCoefficient &vqfc) : vqfc(
```

Comment:
This constructor is not needed as the one below can be called with null for second argument.

Reply:
Deleted now